src/gpu/nodeKernels.cu

```diff
@@ -100,6 +100,117 @@ __device__ Vec3f reflectPolarization(const Vec3f& pol, const Vec3f& hitNormal, c
 	return -polCompR * refPolR + polCompU * refPolU;
 }
 
+__device__ void saveReturnAsHit(const RaytraceRequestContext* ctx, int beamIdx, int sampleIdx, int returnPointIdx,
+                                int returnType)
+{
+	if (ctx->xyz != nullptr) {
+		const Mat3x4f ray = ctx->raysWorld[beamIdx];
+		const Vec3f origin = ray * Vec3f{0, 0, 0};
+		const Vec3f dir = (ray * Vec3f{0, 0, 1} - origin).normalized();
+		ctx->xyz[returnPointIdx] = origin + dir * ctx->mrSamples.distance[sampleIdx];
+	}
+	if (ctx->isHit != nullptr) {
+		ctx->isHit[returnPointIdx] = ctx->mrSamples.isHit[sampleIdx];
+	}
+	if (ctx->ringIdx != nullptr && ctx->ringIds != nullptr) {
+		ctx->ringIdx[returnPointIdx] = ctx->ringIds[beamIdx % ctx->ringIdsCount];
```

Comment:
Probably, `ringIdx` can be filled in `saveBeamSharedData()` regardless of hit or non-hit.

Reply:
I updated code like that.